tools/openremoteslide-quickhash1sum.1
.\"
.\" OpenSlide, a library for reading whole slide image files
.\"
.\" Copyright (c) 2007-2012 Carnegie Mellon University
.\" All rights reserved.
.\"
.\" OpenSlide is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License as
.\" published by the Free Software Foundation, version 2.1.
.\"
.\" OpenSlide is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with OpenSlide. If not, see
.\" <http://www.gnu.org/licenses/>.
.\"


.\" See man-pages(7) for formatting conventions.


.TH OPENSLIDE-QUICKHASH1SUM 1 2012-02-24 "OpenSlide 341.0.1" "User Commands"

.mso www.tmac

.SH NAME
openremoteslide-quickhash1sum \- Print OpenSlide quickhash-1 checksums

.SH SYNOPSIS
.BR "openremoteslide-quickhash1sum " [ --help "] [" --version ]
.IR slide ...

.SH DESCRIPTION
Print OpenSlide
.I quickhash-1
checksums for one or more virtual slide files, in a format similar to
.BR sha256sum (1).

.I quickhash-1
is a non-cryptographic, 256-bit hash of a subset of a slide's data.
It uniquely identifies a particular virtual slide, but cannot be used to
detect corruption or modification of the slide file.

.SH OPTIONS
.TP
.B --help
Display usage summary.

.TP
.B --version
Display version and copyright information.

.SH EXIT STATUS
.B openremoteslide-quickhash1sum
returns 0 on success, 1 if a
.I quickhash-1
could not be calculated, or 2 if the arguments are invalid.

.SH BUGS
.I quickhash-1
is not defined for all slide files supported by OpenSlide.  If a slide
does not have a
.IR quickhash-1 ,
"No quickhash-1 available" will be printed.

.SH COPYRIGHT
Copyright \(co 2007-2012 Carnegie Mellon University and others

OpenSlide is free software: you can redistribute it and/or modify it under
the terms of the
.URL http://gnu.org/licenses/lgpl-2.1.html "GNU Lesser General Public License, version 2.1" .

OpenSlide comes with NO WARRANTY, to the extent permitted by law.  See the
GNU Lesser General Public License for more details.

.SH SEE ALSO
.BR openremoteslide-show-properties (1),
.BR openremoteslide-write-png (1),
.BR sha256sum (1)


CHANGELOG.txt
Notable Changes in OpenSlide
============================

Version 341.0.1, 2015-04-20
 * New formats: Philips TIFF, Ventana TIFF
 * Support OpenJPEG 2.1.0
 * Improve performance of JPEG and JP2K decoding
 * Add openremoteslide.region[i].* properties
 * Improve MATLAB compatibility
 * Enable function deprecation warnings with MSVC
 * Many portability fixes
 * aperio: Detect OpenJPEG chroma subsampling breakage during open
 * aperio: Fill in missing tiles with downsampled data
 * aperio: Report MPP for slides scanned in locales with decimal comma
 * hamamatsu: Support NDPI files > 4 GB
 * hamamatsu: Properly detect NDPI slides produced by NDP.toolkit
 * hamamatsu: Support VMS/VMU slides without a NoLayers key
 * hamamatsu: Report MPP for VMS/VMU
 * leica: Support slides with 2010/03/10 XML namespace
 * leica: Base64-decode leica.barcode property in 2010/10/01 namespace
 * sakura: Support slides with multiple focal planes
 * sakura: Support slides without tile table
 * ventana: Support slides with multiple focal planes
 * ventana: Improve positioning of AOIs within level
 * ventana: Fix failure to recognize macro image on some slides

Version 3.4.0, 2014-01-25
 * Major internal restructuring
 * New formats: Hamamatsu NDPI, Sakura SVSLIDE, Ventana BIF (preliminary)
 * Add openremoteslide_detect_vendor()
 * Deprecate openremoteslide_can_open() (not very useful and often misused)
 * Document performance considerations for openremoteslide_open()
 * Add properties giving the bounds of the non-empty region of the slide
 * leica: Support multiple main images if their levels are coplanar
 * leica: Use slide size as level size
 * mirax: Support PNG- and BMP-formatted slides
 * mirax: Fix "Expected first 0 value" error opening some slides
 * mirax: Fix incorrect tile placement on some slides without overlaps
 * mirax: Never synthesize downsampled levels
 * Add OPENSLIDE_DEBUG environment variable (OPENSLIDE_DEBUG=? for help)
 * Fix some crashes in error paths
 * Add tests for many error paths

Version 3.3.3, 2013-04-13
 * Fix inclusion of openremoteslide.h with MSVC
 * Properly handle Aperio JP2K slides with zero-length tiles
 * Support Hamamatsu slides with blank MacroImage key

Version 3.3.2, 2012-12-01
 * Fix seams in MIRAX 2.2 slides (thanks, Agelos Pappas)
 * Fix associated image naming in single-level Aperio slides
 * Stop decoding MIRAX tiles outside requested region
 * Stop decoding unneeded tiles during tile-aligned accesses
 * Increase Hamamatsu VMU tile size to reduce rendering overhead
 * Document performance considerations for openremoteslide_can_open()

Version 3.3.1, 2012-10-14
 * Parallelize concurrent openremoteslide_read_region calls on an openremoteslide_t
 * Eliminate background scanning of tile headers in MIRAX
 * Scan many fewer tiles during first accesses to Hamamatsu VMS
 * Ignore Leica Z-planes other than 0
 * Add experimental tile-size properties
 * Document API thread safety

Version 3.3.0, 2012-09-08
 * Support Leica SCN format (requires libtiff >= 4) (thanks, Agelos Pappas)
 * Allow opening MIRAX 2.2 slides (though there are seams, bug #92)
 * Add standardized microns-per-pixel and objective-power properties
 * Add "macro" associated image in Trestle
 * Rename "layer" to "level" throughout the API (deprecate "layer" functions;
   remove "layer" properties)
 * Report parse errors in openremoteslide_open() by returning an openremoteslide_t in
   error state
 * Deprecate openremoteslide_get_comment()
 * Add openremoteslide_get_version()
 * Improve command-line tools; add manpages
 * Support building with MinGW-w64; drop CMake, MSVC, mingw32
 * Add tests for many error paths

Version 3.2.6, 2012-02-23
 * Support downsampled MIRAX files
 * Improve performance on MIRAX slides without tile overlaps
 * Fix openremoteslide_read_region for large dimensions on layer > 0
   (3.2.5 regression)
 * Correct subpixel error in MIRAX tile placement
 * Fix unlikely use-after-free with Hamamatsu VMU

Version 3.2.5, 2011-12-16
 * Support MIRAX 1.03 files (thanks, Jan Harkes)
 * Fix openremoteslide_read_region for large dimensions
 * Use subpixel precision in all backends
 * Don't keep associated images in memory
 * Disable quickhash-1 for TIFF files with very large top layer
 * Various build fixes (thanks, Jan, Marco Feuerstein, and
   Mathieu Malaterre)
 * Fix some unlikely memory leaks

Version 3.2.4, 2011-03-07
 * Support MIRAX files that do not have non-hierarchical data
   (thanks, Jan Harkes)
 * Fix compilation on Windows (thanks Hauke Heibel)
 * Work around a bug in GKeyFile parser (thanks, Jan)

Version 3.2.3, 2010-09-09
 * Support MIRAX files that use a variant format for tile
   positions (thanks, Hauke Heibel and Marco Feuerstein)
 * Update location of website
 * Add background color property, for slides that have it
 * Update CMake scripts and other Windows fixes
   (thanks Hauke and Marco)
 * Fix some test.c bugs
 * Fix incorrect MIRAX drawing at certain resolutions
   (thanks Hauke and Marco)
 * Support quickhash-1 on older systems (thanks, Jan Harkes)

Version 3.2.2, 2010-06-16
 * Rework some internals of openremoteslide_read_region
 * Support negative coordinates and zero-sized dimensions
   in openremoteslide_read_region
 * Clarify the documentation about openremoteslide_read_region
 * Fix Windows build bug with new NGR support
 * Enable untested BigTIFF support

Version 3.2.1, 2010-06-03
 * Fix crashes on Windows when trying to read Hamamatsu files
 * Fix jpeg 7 problems in read_from_one_jpeg
 * Quiet the error handling system after the first error

Version 3.2.0, 2010-06-01
 * Add experimental CMake support and fixes for building with MSVC
   (thanks to Hauke Heibel!)
 * Enable detecting runtime errors
 * Add initial Hamamatsu Nanozoomer VMU support
   (thanks to Steve Lamont)
 * Add "openremoteslide-write-png" tool

Version 3.1.1, 2010-04-27
 * Fix a crash when reading an invalid VMS file
 * Fix memory leaks when reading an invalid VMS file

tools/openremoteslide-quickhash1sum.c
/*
 *  OpenSlide, a library for reading whole slide image files
 *
 *  Copyright (c) 2010-2012 Carnegie Mellon University
 *  All rights reserved.
 *
 *  OpenSlide is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation, version 2.1.
 *
 *  OpenSlide is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with OpenSlide. If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <glib.h>
#include "openremoteslide.h"
#include "openremoteslide-tools-common.h"

static gboolean process(const char *file) {
  openremoteslide_t *osr = openremoteslide_open(file);
  if (osr == NULL) {
    fprintf(stderr, "%s: %s: Not a file that OpenSlide can recognize\n",
	    g_get_prgname(), file);
    fflush(stderr);
    return FALSE;
  }

  const char *err = openremoteslide_get_error(osr);
  if (err) {
    fprintf(stderr, "%s: %s: %s\n", g_get_prgname(), file, err);
    fflush(stderr);
    openremoteslide_close(osr);
    return FALSE;
  }

  const char *hash = openremoteslide_get_property_value(osr,
        "openremoteslide.quickhash-1");
  if (hash != NULL) {
    printf("%s  %s\n", hash, file);
  } else {
    fprintf(stderr, "%s: %s: No quickhash-1 available\n", g_get_prgname(),
            file);
    fflush(stderr);
    openremoteslide_close(osr);
    return FALSE;
  }

  openremoteslide_close(osr);
  return TRUE;
}


static const struct openremoteslide_tools_usage_info usage_info = {
  "FILE...",
  "Print OpenSlide quickhash-1 (256-bit) checksums.",
};

int main (int argc, char **argv) {
  _openremoteslide_tools_parse_commandline(&usage_info, &argc, &argv);
  if (argc < 2) {
    _openremoteslide_tools_usage(&usage_info);
  }

  int ret = 0;
  for (int i = 1; i < argc; i++) {
    if (!process(argv[i])) {
      ret = 1;
    }
  }

  return ret;
}
